src/mongoose_listener_config.erl

```diff
@@ -0,0 +1,80 @@
+%% @doc Utilities related to listener configuration options
+
+-module(mongoose_listener_config).
+
+-export([ensure_ip_options/1,
+         verify_unique_listeners/1,
+         prepare_opts/1,
+         filter_socket_opts/1,
+         listener_id/1]).
+
+-type listener() :: #{port := inet:port_number(),
+                      ip_tuple := inet:ip_address(),
+                      ip_address := string(),
+                      ip_version := 4 | 6,
+                      proto := proto(),
+                      any() => any()}.
+-type listener_id() :: {inet:port_number(), inet:ip_address(), proto()}.
+-type proto() :: tcp | udp.
+
+-export_type([listener/0, listener_id/0, proto/0]).
+
+%% @doc Fill in IP-related options that can be calculated automatically.
+%% Apart from these options, the input should be a complete listener configuration.
+-spec ensure_ip_options(map()) -> listener().
+ensure_ip_options(Opts = #{ip_address := IPAddr, ip_version := 4}) ->
+    {ok, IPTuple} = inet:parse_ipv4_address(IPAddr),
+    Opts#{ip_tuple => IPTuple};
+ensure_ip_options(Opts = #{ip_address := IPAddr, ip_version := 6}) ->
+    {ok, IPTuple} = inet:parse_ipv6_address(IPAddr),
+    Opts#{ip_tuple => IPTuple};
+ensure_ip_options(Opts = #{ip_address := IPAddr}) ->
+    {ok, IPTuple} = inet:parse_address(IPAddr),
+    Opts#{ip_tuple => IPTuple,
+          ip_version => ip_version(IPTuple)};
+ensure_ip_options(Opts = #{ip_version := 6}) ->
+    ensure_ip_options(Opts#{ip_address => "::"});
+ensure_ip_options(Opts) ->
+    ensure_ip_options(Opts#{ip_address => "0"}).
+
+ip_version(T) when size(T) =:= 4 -> 4;
+ip_version(T) when size(T) =:= 8 -> 6.
```

Comment:
Might as well use `tuple_size/1`, doesn't really matter much in this specific case but I think it's a better practice, it's "theoretically faster" (doesn't need to select the sizing function based on the type but instead just applies or fails), and dialyzer/gradualizer can return better diagnosis. All in theory 😄 
```suggestion
ip_version(T) when tuple_size(T) =:= 4 -> 4;
ip_version(T) when tuple_size(T) =:= 8 -> 6.
```

Reply:
Yes it's better to match on the proper type, I just copy-pasted this ;)